cirq-core/cirq/ops/classically_controlled_operation.py

```diff
@@ -169,12 +169,31 @@ def not_zero(measurement):
     def _with_measurement_key_mapping_(
         self, key_map: Dict[str, str]
     ) -> 'ClassicallyControlledOperation':
-        keys = [protocols.with_measurement_key_mapping(k, key_map) for k in self._control_keys]
-        return self._sub_operation.with_classical_controls(*keys)
+        return ClassicallyControlledOperation(
+            self._sub_operation,
+            [protocols.with_measurement_key_mapping(k, key_map) for k in self._control_keys],
+        )
 
-    def _with_key_path_prefix_(self, path: Tuple[str, ...]) -> 'ClassicallyControlledOperation':
-        keys = [protocols.with_key_path_prefix(k, path) for k in self._control_keys]
-        return self._sub_operation.with_classical_controls(*keys)
+    def _with_key_path_prefix_(
+        self,
+        path: Tuple[str, ...],
+        local_keys: FrozenSet[value.MeasurementKey],
+        extern_keys: FrozenSet[value.MeasurementKey],
+    ) -> 'ClassicallyControlledOperation':
+        def map_key(key: value.MeasurementKey) -> value.MeasurementKey:
+            if key in local_keys:
+                return protocols.with_key_path_prefix(key, path)
+            for i in range(len(path)):
+                back_path = path[0 : len(path) - i]
+                new_key = protocols.with_key_path_prefix(key, back_path)
+                if new_key in extern_keys:
+                    return new_key
```

Comment:
Yeah, this behavior is really peculiar for a method named `_with_key_path_prefix_`. I'm in support of a new `with_rescoped_key` protocol and (if possible) deprecating `with_key_path_prefix`.